shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,388 @@
+  if (self = [super init]) {
+    _texture = texture;
+    _surface = surface;
+  }
+  return self;
+}
+
+@end
+
+@interface FlutterDrawable : NSObject <CAMetalDrawable> {
+  FlutterTexture* _texture;
+  __weak FlutterMetalLayer* _layer;
+  NSUInteger _drawableId;
+  BOOL _presented;
+}
+
+- (instancetype)initWithTexture:(FlutterTexture*)texture
+                          layer:(FlutterMetalLayer*)layer
+                     drawableId:(NSUInteger)drawableId;
+
+@end
+
+@implementation FlutterDrawable
+
+- (instancetype)initWithTexture:(FlutterTexture*)texture
+                          layer:(FlutterMetalLayer*)layer
+                     drawableId:(NSUInteger)drawableId {
+  if (self = [super init]) {
+    _texture = texture;
+    _layer = layer;
+    _drawableId = drawableId;
+  }
+  return self;
+}
+
+- (id<MTLTexture>)texture {
+  return self->_texture.texture;
+}
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+- (CAMetalLayer*)layer {
+  return (id)self->_layer;
+}
+#pragma clang diagnostic pop
+
+- (NSUInteger)drawableID {
+  return self->_drawableId;
+}
+
+- (CFTimeInterval)presentedTime {
+  return 0;
+}
+
+- (void)present {
+  [_layer presentTexture:self->_texture];
+  self->_presented = YES;
+}
+
+- (void)dealloc {
+  if (!_presented) {
+    [_layer returnTexture:self->_texture];
+  }
+}
+
+- (void)addPresentedHandler:(nonnull MTLDrawablePresentedHandler)block {
+}
+
+- (void)presentAtTime:(CFTimeInterval)presentationTime {
+}
+
+- (void)presentAfterMinimumDuration:(CFTimeInterval)duration {
+}
+
+@end
+
+@implementation FlutterMetalLayer
+
+@synthesize preferredDevice = _preferredDevice;
+@synthesize device = _device;
+@synthesize pixelFormat = _pixelFormat;
+@synthesize framebufferOnly = _framebufferOnly;
+@synthesize colorspace = _colorspace;
+@synthesize wantsExtendedDynamicRangeContent = _wantsExtendedDynamicRangeContent;
+
+- (instancetype)init {
+  if (self = [super init]) {
+    _preferredDevice = MTLCreateSystemDefaultDevice();
+    self.device = self.preferredDevice;
+    self.pixelFormat = MTLPixelFormatBGRA8Unorm;
+    _availableTextures = [[NSMutableSet alloc] init];
+
+    _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(onDisplayLink:)];
+    [self setMaxRefreshRate:[DisplayLinkManager displayRefreshRate] forceMax:NO];
+    [_displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(didEnterBackground:)
+                                                 name:UIApplicationDidEnterBackgroundNotification
+                                               object:nil];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)setMaxRefreshRate:(double)refreshRate forceMax:(BOOL)forceMax {
```

Comment:
Our display link usage is a bit of a mess. I'd like to clean it up at some point, hopefully also fixing https://github.com/flutter/flutter/issues/110431 in the process.

As far as this PR is concerned, the main issue is that the iOS vsync waiter is scheduling the display link on UI thread, which does not trigger a core animation frame (`CAMetalLayer` does not seem to care). So the vsync waiter displaylink is running at 120hz while core animation only syncs to compositor at 60hz. And because I don't want to touch other parts in this PR, I added a display link and copied some of the logic from vsync waiter so that it runs at same speed.